### PR TITLE
feat: Add new evaluateAsync function

### PR DESCRIPTION
**Pull request by JimLiu**
```js
browser.evaluateAsync(3, 6, function(a, b) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(a * b);
      }, 1000);
    });
  })
```

image

### CODE REVIEW

**Comment by dbushong** [comment]
Would it be possible to detect that the arity of the passed function is 1 more than the passed args, and if so, do the callbacky version, else wrap it in a promise resolution?

Honestly since this is the promise-chain driver, i'd be ok with *only* supporting promises.

**Comment by jkrems** [comment]
👍 for making this promise only. Supporting a callback in the middle of promise code seems dangerous.

**Comment by JimLiu** [comment]
@dbushong @jkrems Do you have a sample code for it? Couldn't get it works with Promise.

File: test/integration/evaluate.test.js [hunk 1]
```diff
@@ -41,4 +44,24 @@ describe('evaluate', () => {
       );
     })
   );
+
+  it(
+    'runs JavaScript passed as an Async Function with callback',
+    co.wrap(function*() {
+      if (browserName !== 'chrome') {
+        //They only work for chrome.
```
Comment (dbushong) [inline test/integration/evaluate.test.js]: maybe also firefox?  could we do some light feature-detection here?  what is it that doesn't work on some browsers?

**Comment by dbushong** [comment]
re: promise: i'll give it a stab

**Comment by jkrems** [comment]
If `wd` only supports callbacks for the function being run, maybe we can wrap it?

```js
// pseudo-code
function wrap(actual, ...args, callback) {
  new Promise(resolve => resolve(actual(...args))).then(data => callback(null, data), callback);
}
```

File: lib/browser/evaluate.js [hunk 2]
```diff
@@ -56,3 +56,26 @@ exports.evaluate = function evaluate() {
...
+  const invocation =
+    'evaluateAsync(clientFunction) - requires (Function|String) clientFunction';
+  assert.truthy(invocation, clientFunction);
+
+  switch (typeof clientFunction) {
+    case 'function':
+      clientFunction = `var done = arguments[arguments.length-1], args=[].slice.call(arguments, 0, arguments.length-1);
+      return new Promise(
+        (resolve) => resolve((${clientFunction}).apply(this, args))
+      ).then(done)`;
```
Comment (dbushong) [inline lib/browser/evaluate.js]: Do we need the `new Promise()` here?  Maybe if we want to make sure it's a Promise, just:

```js
        return Promise.resolve((${clientFunction}).apply(this, args)).then(done)
```

Comment (jkrems) [reply lib/browser/evaluate.js]: It depends on the behavior we want for uncaught/sync errors. Errors thrown within the promise constructor are turned into rejections.

File: lib/browser/evaluate.js [hunk 3]
```diff
@@ -56,3 +56,22 @@ exports.evaluate = function evaluate() {
...
+  assert.truthy(invocation, clientFunction);
+
+  if (typeof clientFunction !== 'function') {
+    throw new Error(invocation);
+  }
+
+  clientFunction = `var done = arguments[arguments.length-1], args=[].slice.call(arguments, 0, arguments.length-1);
+      return new Promise(
+        (resolve) => resolve((${clientFunction}).apply(this, args))
+      ).then(done)`;
```
Comment (jkrems) [inline lib/browser/evaluate.js]: The `.then(done)` means that rejections just make this hang (unless I'm missing something) - can we make sure that we have a test case with an error?

Comment (dbushong) [reply lib/browser/evaluate.js]: good point - contrary to the docs, the `done()` doesn't accept error-first-arg :-(

any thoughts on a good way to get the errors out?

Comment (jkrems) [reply lib/browser/evaluate.js]: We could wrap this in `{ reason, value }` and convert to rejection in the API method?

File: test/integration/evaluate.test.js [hunk 4]
```diff
@@ -42,3 +45,70 @@ describe('evaluate', () => {
     })
   );
 });
+
+describe('evaluateAsync', () => {
+  if (browserName === 'phantomjs') {
+    xit("skipping tests because browser phantomjs doesn't support alerts");
```
Comment (dbushong) [inline test/integration/evaluate.test.js]: ...doesn't support evaluateAsync

File: test/integration/evaluate.test.js [hunk 5]
```diff
@@ -42,3 +45,70 @@ describe('evaluate', () => {
...
+  before(browser.beforeHook());
+
+  before(() => browser.loadPage('/'));
+
+  it(
+    'runs JavaScript passed as a Function',
+    co.wrap(function*() {
+      assert.equal(
+        6,
+        // eslint-disable-next-line prefer-arrow-callback
```
Comment (dbushong) [inline test/integration/evaluate.test.js]: if we're skipping phantom anyway, we can just go ahead and use arrow functions for all these

File: lib/browser/evaluate.js [hunk 6]
```diff
@@ -56,3 +56,22 @@ exports.evaluate = function evaluate() {
...
+
+  const invocation =
+    'evaluateAsync(clientFunction) - requires (Function) clientFunction';
+  assert.truthy(invocation, clientFunction);
+
+  if (typeof clientFunction !== 'function') {
+    throw new Error(invocation);
+  }
+
+  clientFunction = `var done = arguments[arguments.length-1], args=[].slice.call(arguments, 0, arguments.length-1);
```
Comment (dbushong) [inline lib/browser/evaluate.js]: simpler:
`var args = [].slice.call(arguments, 0), done = args.pop();`

File: lib/browser/evaluate.js [hunk 7]
```diff
@@ -56,3 +56,33 @@ exports.evaluate = function evaluate() {
...
+
+  clientFunction = `var args = [].slice.call(arguments), done = args.pop();
+    try {
+      var result = (${clientFunction}).apply(this, args);
+      Promise.resolve(result)
+      .then(result => done({reason: null, value: result}))
+      .catch(err => done({reason: JSON.stringify(err)}));
+    } catch(err) {
+      done({reason: JSON.stringify(err)})
+    }`;
```
Comment (dbushong) [inline lib/browser/evaluate.js]: i think you can skip the try/catch:
```js
Promise
  .resolve()
  .then(() => (${clientFunction}).apply(this, args))
  .then(
    value => done({ value }),
    err => done({ reason: err && { message: err.message, stack: err.stack } })
  );
```

File: lib/browser/evaluate.js [hunk 8]
```diff
@@ -56,3 +56,33 @@ exports.evaluate = function evaluate() {
...
+
+  return this.executeAsync(clientFunction, args).then(result => {
+    if (result.reason) {
+      throw JSON.parse(result.reason);
+    }
```
Comment (dbushong) [inline lib/browser/evaluate.js]: the response will already be JSON-encoded, so assuming you do something as shown above, you could then rehydrate it with something like:
```js
.then(({ reason, value }) => {
  if (reason) throw Object.assign(new Error(reason.message), reason);
  return value;
})

**Comment by JimLiu** [comment]
Updated, thanks!

**Pull request merged**